File: packages/commerce-types/tests/multi-agent.test.ts
import { describe, expect, it } from "vitest";

import { createMultiAgentSession } from "../src/multi-agent.js";
import { commitmentVersion } from "../src/guard.js";
import { newCommitment, partyId, valueId, type Value } from "../src/primitives.js";
import { applyCommitmentPath } from "../src/transitions.js";

const buyer = partyId("buyer");
const seller = partyId("seller");

function moneyValue(amount: number): Value {
  return { id: valueId(), form: { kind: "Money", money: { amount, currency: "MAD" } }, quantity: 1, state: { type: "Available" } };
}

function fulfilledOrder(amount: number) {
  return applyCommitmentPath(newCommitment(buyer, seller, { offered: [], requested: [moneyValue(amount)] }), { type: "Fulfilled" }, seller);
}

function refund(commitment: string, amount: number, actor: string, key: string) {
  return { commitment, to: { type: "Refunded" as const, amount: { amount, currency: "MAD" }, at: "2026-02-01T00:00:00.000Z" }, actor, idempotencyKey: key };
}

describe("createMultiAgentSession — shared-world enforcement with attribution", () => {
  it("catches a cumulative violation across DIFFERENT actors, attributed to the tipping actor", () => {
    const order = fulfilledOrder(200);
    const id = String(order.id);
    const session = createMultiAgentSession({ commitments: [order], fulfillments: [], parties: [] });

    expect(session.propose(refund(id, 80, "agent-A", "a")).ok).toBe(true);
    expect(session.propose(refund(id, 80, "agent-B", "b")).ok).toBe(true);
    const third = session.propose(refund(id, 80, "agent-C", "c"));
    expect(third.ok).toBe(false);
    if (!third.ok) {
      expect(third.violations[0]?.rule).toBe("I-1");
      // attribution names the tipping actor (agent-C), applied after the others.
      expect(third.actor).toBe("agent-C");
      expect(third.attribution).toContain("agent-C");
      expect(third.attribution).toContain("agent-A");
      expect(third.attribution).toContain("agent-B");
      expect(third.attribution).not.toContain("conspir"); // NOT collusion
    }
    // the world did not advance past the two accepted refunds.
    expect(session.refundedSoFar(id)?.amount).toBe(160);
  });

  it("attributes to the right actor regardless of which one tips it", () => {
    const order = fulfilledOrder(200);
    const id = String(order.id);
    const session = createMultiAgentSession({ commitments: [order], fulfillments: [], parties: [] });
    session.propose(refund(id, 120, "finance-agent", "f"));
    const tip = session.propose(refund(id, 100, "support-agent", "s")); // 220 > 200
    expect(tip.ok).toBe(false);
    if (!tip.ok) {
      expect(tip.actor).toBe("support-agent");
      expect(tip.attribution).toContain("finance-agent"); // the accumulated context
    }
  });

  it("a single-actor session behaves identically (attribution is additive)", () => {
    const order = fulfilledOrder(200);
    const id = String(order.id);
    const session = createMultiAgentSession({ commitments: [order], fulfillments: [], parties: [] });
    // Same actor does all three — still caught at the third (attribution just names that actor).
    expect(session.propose(refund(id, 80, "solo", "a")).ok).toBe(true);
    expect(session.propose(refund(id, 80, "solo", "b")).ok).toBe(true);
    const third = session.propose(refund(id, 80, "solo", "c"));
    expect(third.ok).toBe(false);
    if (!third.ok) {
      expect(third.actor).toBe("solo");
      expect(third.attribution).toContain("no prior actions"); // no OTHER actors
    }
  });

  it("a valid multi-agent sequence completes, with a per-actor summary", () => {
    const draft = newCommitment(buyer, seller);
    const id = String(draft.id);
    const session = createMultiAgentSession({ commitments: [draft], fulfillments: [], parties: [] });
    expect(session.propose({ commitment: id, to: { type: "Proposed" }, actor: "buyer-agent" }).ok).toBe(true);
    expect(session.propose({ commitment: id, to: { type: "Accepted" }, actor: "seller-agent" }).ok).toBe(true);
    expect(session.propose({ commitment: id, to: { type: "Active" }, actor: "ops-agent" }).ok).toBe(true);
    expect(session.world.commitments[0]?.state.type).toBe("Active");
    expect(session.actorsSummary()).toEqual({ "buyer-agent": 1, "seller-agent": 1, "ops-agent": 1 });
    expect(session.log.map((r) => r.actor)).toEqual(["buyer-agent", "seller-agent", "ops-agent"]);
  });

  it("an F3 stale-version conflict between two agents is attributed to the late actor", () => {
    const order = applyCommitmentPath(newCommitment(buyer, seller), { type: "Accepted" }, seller);
    const id = String(order.id);
    const session = createMultiAgentSession({ commitments: [order], fulfillments: [], parties: [] });
    const planned = commitmentVersion(session.world.commitments[0]!);

    // agent-A advances the commitment.
    expect(session.propose({ commitment: id, to: { type: "Active" }, actor: "agent-A", expectedVersion: planned, idempotencyKey: "A" }).ok).toBe(true);
    // agent-B planned against the stale version → conflict, attributed to agent-B.
    const b = session.propose({ commitment: id, to: { type: "Disputed", by: buyer, reason: "x", opened_at: "2026-03-01T00:00:00.000Z" }, actor: "agent-B", expectedVersion: planned, idempotencyKey: "B" });
    expect(b.ok).toBe(false);
    if (!b.ok) {
      expect(b.conflict).toBe(true);
      expect(b.actor).toBe("agent-B");
      expect(b.attribution).toContain("conflict");
    }
  });

  it("an F4 replay by the same actor is a replay and is not double-logged", () => {
    const order = fulfilledOrder(200);
    const id = String(order.id);
    const session = createMultiAgentSession({ commitments: [order], fulfillments: [], parties: [] });
    const first = session.propose(refund(id, 50, "agent-A", "k"));
    expect(first.ok).toBe(true);
    const retry = session.propose(refund(id, 50, "agent-A", "k"));
    expect(retry.ok).toBe(true);
    if (retry.ok) expect(retry.replay).toBe(true);
    // logged once (the replay applied nothing new); refunded once.
    expect(session.actorsSummary()).toEqual({ "agent-A": 1 });
    expect(session.refundedSoFar(id)?.amount).toBe(50);
  });
});

File: bindings/go/multi_agent_test.go
package warp

import (
	"encoding/json"
	"fmt"
	"strings"
	"testing"

	gen "github.com/yasirlts/warp-lang/bindings/go/generated"
)

func acceptedOrderMA(id string) gen.Commitment {
	var c gen.Commitment
	j := fmt.Sprintf(`{"id":"%s","parties":{"initiator":"buyer","counterparty":"seller","intermediaries":[]},"subject":{"offered":[],"requested":[{"id":"v","form":{"kind":"Money","money":{"amount":200,"currency":"MAD"}},"quantity":1,"state":{"type":"Available"}}]},"state":{"type":"Accepted"},"history":[],"children":[],"created_at":"2026-01-02T08:00:00.000Z"}`, id)
	if err := json.Unmarshal([]byte(j), &c); err != nil {
		panic(err)
	}
	return c
}

func draftOrderMA(id string) gen.Commitment {
	var c gen.Commitment
	j := fmt.Sprintf(`{"id":"%s","parties":{"initiator":"buyer","counterparty":"seller","intermediaries":[]},"subject":{"offered":[],"requested":[]},"state":{"type":"Draft"},"history":[],"children":[],"created_at":"2026-01-02T08:00:00.000Z"}`, id)
	if err := json.Unmarshal([]byte(j), &c); err != nil {
		panic(err)
	}
	return c
}

func disputedStateMA() gen.CommitmentState {
	var s gen.CommitmentState
	_ = json.Unmarshal([]byte(`{"type":"Disputed","by":"buyer","reason":"x","opened_at":"2026-03-01T00:00:00.000Z"}`), &s)
	return s
}

func refundByActor(id string, amount float64, actor, key string) ProposedAction {
	return ProposedAction{Commitment: id, To: refundState(amount), Actor: actor, IdempotencyKey: key}
}

func TestMACatchesCumulativeViolationAcrossActorsAttributed(t *testing.T) {
	s := CreateMultiAgentSession(World{Commitments: []gen.Commitment{fulfilledOrder("order_1", 200)}})
	if !s.Propose(refundByActor("order_1", 80, "agent-A", "a")).OK {
		t.Fatal("first refund should be accepted")
	}
	if !s.Propose(refundByActor("order_1", 80, "agent-B", "b")).OK {
		t.Fatal("second refund should be accepted")
	}
	third := s.Propose(refundByActor("order_1", 80, "agent-C", "c"))
	if third.OK {
		t.Fatal("expected rejection on the third refund")
	}
	if third.Violations[0].Rule != "I-1" {
		t.Fatalf("expected I-1, got %s", third.Violations[0].Rule)
	}
	if third.Actor != "agent-C" {
		t.Fatalf("expected tipping actor agent-C, got %s", third.Actor)
	}
	for _, want := range []string{"agent-C", "agent-A", "agent-B"} {
		if !strings.Contains(third.Attribution, want) {
			t.Fatalf("attribution %q missing %q", third.Attribution, want)
		}
	}
	if strings.Contains(third.Attribution, "conspir") {
		t.Fatal("attribution must not claim conspiracy/collusion")
	}
	// World did not advance past the two accepted refunds.
	sofar, _, _ := s.RefundedSoFar("order_1")
	if sofar != 160 {
		t.Fatalf("refunded so far = %v, want 160", sofar)
	}
}

func TestMASingleActorBehavesIdenticallyAttributionAdditive(t *testing.T) {
	s := CreateMultiAgentSession(World{Commitments: []gen.Commitment{fulfilledOrder("order_1", 200)}})
	s.Propose(refundByActor("order_1", 80, "solo", "a"))
	s.Propose(refundByActor("order_1", 80, "solo", "b"))
	third := s.Propose(refundByActor("order_1", 80, "solo", "c"))
	if third.OK {
		t.Fatal("expected rejection")
	}
	if third.Actor != "solo" {
		t.Fatalf("expected actor solo, got %s", third.Actor)
	}
	if !strings.Contains(third.Attribution, "no prior actions") {
		t.Fatalf("attribution %q should note no OTHER prior actors", third.Attribution)
	}
}

func TestMAValidSequenceCompletesWithSummaryAndLog(t *testing.T) {
	s := CreateMultiAgentSession(World{Commitments: []gen.Commitment{draftOrderMA("order_1")}})
	if !s.Propose(ProposedAction{Commitment: "order_1", To: gen.CommitmentState{Type: "Proposed"}, Actor: "buyer-agent"}).OK {
		t.Fatal("propose Proposed should be accepted")
	}
	if !s.Propose(ProposedAction{Commitment: "order_1", To: gen.CommitmentState{Type: "Accepted"}, Actor: "seller-agent"}).OK {
		t.Fatal("propose Accepted should be accepted")
	}
	if !s.Propose(ProposedAction{Commitment: "order_1", To: gen.CommitmentState{Type: "Active"}, Actor: "ops-agent"}).OK {
		t.Fatal("propose Active should be accepted")
	}
	if got := s.World().Commitments[0].State.Type; got != "Active" {
		t.Fatalf("final state = %s, want Active", got)
	}
	summary := s.ActorsSummary()
	for _, actor := range []string{"buyer-agent", "seller-agent", "ops-agent"} {
		if summary[actor] != 1 {
			t.Fatalf("actor %s count = %d, want 1", actor, summary[actor])
		}
	}
	wantOrder := []string{"buyer-agent", "seller-agent", "ops-agent"}
	for i, r := range s.Log() {
		if r.Actor != wantOrder[i] {
			t.Fatalf("log[%d] actor = %s, want %s", i, r.Actor, wantOrder[i])
		}
	}
}

func TestMAStaleVersionConflictAttributedToLateActor(t *testing.T) {
	s := CreateMultiAgentSession(World{Commitments: []gen.Commitment{acceptedOrderMA("order_1")}})
	planned := CommitmentVersion(&s.World().Commitments[0])

	a := ProposedAction{Commitment: "order_1", To: gen.CommitmentState{Type: "Active"}, Actor: "agent-A", ExpectedVersion: planned, IdempotencyKey: "A"}
	if !s.Propose(a).OK {
		t.Fatal("agent-A activate should be accepted")
	}
	b := ProposedAction{Commitment: "order_1", To: disputedStateMA(), Actor: "agent-B", ExpectedVersion: planned, IdempotencyKey: "B"}
	verdict := s.Propose(b)
	if !verdict.Conflict {
		t.Fatal("expected a conflict on the stale plan")
	}
	if verdict.Actor != "agent-B" {
		t.Fatalf("expected actor agent-B, got %s", verdict.Actor)
	}
	if !strings.Contains(verdict.Attribution, "conflict") {
		t.Fatalf("attribution %q should mention conflict", verdict.Attribution)
	}
}

func TestMAReplayBySameActorIsReplayAndNotDoubleLogged(t *testing.T) {
	s := CreateMultiAgentSession(World{Commitments: []gen.Commitment{fulfilledOrder("order_1", 200)}})
	first := s.Propose(refundByActor("order_1", 50, "agent-A", "k"))
	if !first.OK || first.Replay {
		t.Fatal("first refund should be accepted and not a replay")
	}
	retry := s.Propose(refundByActor("order_1", 50, "agent-A", "k"))
	if !retry.OK || !retry.Replay {
		t.Fatal("retry with same key should be an accepted replay")
	}
	if s.ActorsSummary()["agent-A"] != 1 {
		t.Fatalf("agent-A logged %d times, want 1 (replay applied nothing new)", s.ActorsSummary()["agent-A"])
	}
	sofar, _, _ := s.RefundedSoFar("order_1")
	if sofar != 50 {
		t.Fatalf("refunded so far = %v, want 50 (no double refund)", sofar)
	}
}
